dashboard/app/controllers/openai_chat_controller.rb

```diff
@@ -39,7 +49,40 @@ def validated_level?
     params[:type].present? && params[:type] == 'validation'
   end
 
-  def get_validated_level_test_file_contents(level_id)
+  def add_content_to_system_prompt(system_prompt, level_instructions, test_file_contents)
+    if level_instructions.present?
+      system_prompt += "\n Here are the student instructions for this level: #{level_instructions}"
+    end
+
+    if test_file_contents.present?
+      system_prompt += "\n The contents of the test file are: #{test_file_contents}"
+    end
+
+    system_prompt
+  end
+
+  private def prepend_system_prompt(system_prompt, messages)
+    system_prompt_message = {
+      content: system_prompt,
+      role: "system"
+    }
+
+    messages.unshift(system_prompt_message)
+    messages
+  end
+
+  private def read_file_from_s3(key_path)
+    if [:development, :test].include?(rack_env)
+      local_path = File.join("local-aws", S3_AI_BUCKET, key_path)
+      if File.exist?(local_path)
+        puts "Note: Reading AI prompt from local file: #{key_path}"
+        return File.read(local_path)
+      end
+    end
+    s3_client.get_object(bucket: S3_AI_BUCKET, key: key_path).body.read
```

Comment:
will this cache the response? I'm guessing the system prompt won't change that much?

Reply:
There's a thread relevant to this comment and the one above here: https://codedotorg.slack.com/archives/C051P2V2RN0/p1715274926003069?thread_ts=1715274404.871189&cid=C051P2V2RN0

I copied the dev/test logic from Teacher Tools. I originally had it cached and then disabled it because I was troubleshooting unrelated test failures that mentioned a cache, and Dave said it shouldn't be pilot-blocking. If it's okay with you, I'll reenable the caching, assuming it passes Drone. No strong feelings on the dev/test thing -- though it probably makes sense to disable it for test if we don't know the application :D